main.go

```diff
@@ -0,0 +1,186 @@
+	reflectedServiceType := reflect.ValueOf(serviceType).Elem()
+
+	for _, parameterName := range parameterHierarchy {
+		parameterName = strcase.ToCamel(parameterName)
+
+		if !reflectedServiceType.FieldByName(parameterName).IsValid() {
+			originalParameterName := parameterName
+			var err error
+			parameterName, err = fixStringCase(structs.Names(reflectedServiceType.Interface()), parameterName)
+			if err != nil {
+				err = fmt.Errorf("%w\n%s contains field with name %s : %t",
+					err,
+					reflectedServiceType.Type().Name(),
+					parameterName,
+					reflectedServiceType.FieldByName(parameterName).IsValid(),
+				)
+				return nil, err
+			}
+			info := fmt.Sprintf(
+				"Converted input parameter %s to %s to match correct fieldName\n",
+				originalParameterName,
+				parameterName,
+			)
+			logrus.Info(info)
+		}
+
+		if reflectedServiceType.FieldByName(parameterName).Kind() == reflect.Struct {
+			reflectedServiceType = reflectedServiceType.FieldByName(parameterName)
+		} else if reflectedServiceType.FieldByName(parameterName).Kind() == reflect.String {
+			reflectedServiceType = reflectedServiceType.FieldByName(parameterName)
+			reflectedServiceType.SetString(value)
+			fmt.Printf("setting\n-----------------------\nfield: %s\nvalue: %s\n-----------------------\n", strings.Join(parameterHierarchy, "."), value)
+		} else if reflectedServiceType.FieldByName(parameterName).Kind() == reflect.Int {
+			reflectedServiceType = reflectedServiceType.FieldByName(parameterName)
+			intValue, _ := strconv.Atoi(value)
+			reflectedServiceType.SetInt(int64(intValue))
+			fmt.Printf("setting\n-----------------------\nfield: %s\nvalue: %s\n-----------------------\n", strings.Join(parameterHierarchy, "."), value)
+		} else {
+			err := fmt.Errorf(
+				"cannot assign value %s to field %s with field Type %s",
+				value,
+				strings.Join(parameterHierarchy, "."),
+				reflectedServiceType.FieldByName(parameterName).Kind(),
+			)
+			err = fmt.Errorf("%w\n%s contains field with name %s : %t",
+				err,
+				reflectedServiceType.Type().Name(),
+				parameterName,
+				reflectedServiceType.FieldByName(parameterName).IsValid(),
+			)
+			return nil, err
+		}
+	}
+	return serviceType, nil
+}
+
+// Checks for all field names of a struct if any match the parameter name under Unicode case-folding
+// returns the correct field name if successfull, nil if unsucsessful
+func fixStringCase(fieldNames []string, parameterName string) (string, error) {
+	for _, fieldName := range fieldNames {
+		if strings.EqualFold(fieldName, parameterName) {
+			return fieldName, nil
+		}
+	}
+	return parameterName, fmt.Errorf("could not find field with name %s", parameterName)
+}
+
+// Selects the appropriate service type and initializes it with default values
+func findServiceType(parsedType string) (runtime.Object, error) {
+	parsedType = strings.ToLower(parsedType)
+	var serviceType runtime.Object
+	if strings.Contains(parsedType, "vshn") {
```

Comment:
Maybe one big `switch` statement is enough, something like

```go
switch parsedType {
case "vshnpostgresql":
	serviceType = defaults.GetVSHNPostgreSQLDefault()
case "vshnredis":
	serviceType = defaults.GetVSHNRedisDefault()
// ...
}
```

Or bonus for storing all possible types in a `map[string]func() runtime.Object`, that would also allow to print all known types

```go
const TYPES := map[string]func() runtime.Object{
	"vshnpostgresql": defaults.GetVSHNPostgreSQLDefault,
	"vshnredis": defaults.GetVSHNRedisDefault,
}

// ...

fn, ok := TYPES[parsedType]
if !ok {
	return nil, fmt.Errorf(/* ... */)
}
// ...
```

Reply:
I refactored the defaults to offer the string map of functions.
I think the separation of different providers and with that the switch(although a more lean one) in findServiceType() are easier to maintain and keep track off instead of one big map or switch.